dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageChecker.java

```diff
@@ -182,6 +187,90 @@ private static void printStaticLinkageErrors(
     return jarPaths.build();
   }
 
+  /**
+   * Finds linkage errors in the input classpath and generates a static linkage check report.
+   */
+  StaticLinkageCheckReport findLinkageErrors() {
+    ImmutableList<Path> jarFilePaths = classDumper.getInputClasspath();
+
+    ImmutableMap.Builder<Path, SymbolTable> jarToSymbols = ImmutableMap.builder();
+    for (Path jarPath : jarFilePaths) {
+      jarToSymbols.put(jarPath, scanExternalSymbolTable(jarPath));
+    }
+
+    // Validate linkage error of each reference
+    ImmutableList.Builder<JarLinkageReport> jarLinkageReports = ImmutableList.builder();
+    for (Map.Entry<Path, SymbolTable> entry : jarToSymbols.build().entrySet()) {
+      jarLinkageReports.add(findInvalidReferences(entry.getKey(), entry.getValue()));
+    }
+
+    if (reportOnlyReachable) {
+      // TODO: Optionally, report errors only reachable from entry point classes
+    }
+
+    return StaticLinkageCheckReport.create(jarLinkageReports.build());
+  }
+
+  private JarLinkageReport findInvalidReferences(Path jarPath, SymbolTable jarSymbolTable) {
+    JarLinkageReport.Builder reportBuilder = JarLinkageReport.builder().setJarPath(jarPath);
+
+    // TODO(suztomo): implement validation for field, method and class references in the table
+    return reportBuilder.build();
+  }
+
+  /**
+   * Scans class files in the jar file and creates {@link SymbolTable} populated with class names
+   * defined within the jar and symbolic references to other classes outside the jar.
+   *
+   * @param jarFilePath absolute path to a jar file.
+   * @return Table containing symbol references and classes defined in the jar file.
+   */
+  static SymbolTable scanExternalSymbolTable(Path jarFilePath) {
+    SymbolTable.Builder symbolTableBuilder = SymbolTable.builder();
+    ImmutableSet.Builder<String> symbolTableClassNameBuilder = ImmutableSet.builder();
```

Comment:
If you declare `SymbolTable.Builder` to have methods that return these builders, then you don't even need these variables. Below, you can do things like `symbolTableBuilder.classReferencesBuilder().addAll(partialSymbolTable.getClassReferences())`.